contracts/secret-storage/SecretManager.sol

```diff
@@ -0,0 +1,601 @@
+        uint256 sizeLimit,
+        uint256 endTimestamp,
+        uint256 usdcDeposit,
+        address[] selectedEnclaves
+    );
+
+    event SecretStoreAcknowledgementSuccess(
+        uint256 indexed secretId,
+        address indexed enclaveAddress
+    );
+
+    event SecretStoreAcknowledgementFailed(
+        uint256 indexed secretId
+    );
+
+    event SecretStoreAlive(
+        address indexed enclaveAddress
+    );
+
+    event SecretStoreDead(
+        uint256 indexed secretId,
+        address indexed prevEnclaveAddress,
+        address indexed newEnclaveAddress
+    );
+
+    event SecretResourceUnavailable(
+        uint256 indexed secretId
+    );
+
+    event SecretEndTimestampUpdated(
+        uint256 indexed secretId,
+        uint256 endTimestamp
+    );
+
+    event SecretTerminated(
+        uint256 indexed secretId
+    );
+
+    error SecretManagerInsufficientUsdcDeposit();
+    error SecretManagerInvalidSizeLimit();
+    error SecretManagerInvalidEndTimestamp();
+    error SecretManagerUnavailableResources();
+    error SecretManagerAcknowledgementTimeOver();
+    error SecretManagerAcknowledgementTimeoutPending(address enclaveAddress);
+    error SecretManagerAcknowledgedAlready();
+    error SecretManagerMarkAliveTimeoutOver();
+    error SecretManagerUnacknowledged();
+    error SecretManagerEnclaveNotFound();
+    error SecretManagerStoreIsAlive();
+    error SecretManagerNotUserStoreOwner();
+    error SecretManagerAlreadyTerminated();
+
+    //-------------------------------- internal functions start ----------------------------------//
+
+    function _createSecret(
+        uint256 _sizeLimit,
+        uint256 _endTimestamp,
+        uint256 _usdcDeposit,
+        address _owner
+    ) internal {
+        if(_sizeLimit == 0 || _sizeLimit > GLOBAL_MAX_STORE_SIZE)
+            revert SecretManagerInvalidSizeLimit();
+
+        if ((_endTimestamp < block.timestamp + GLOBAL_MIN_STORE_DURATION) || (_endTimestamp > block.timestamp + GLOBAL_MAX_STORE_DURATION)) 
+            revert SecretManagerInvalidEndTimestamp();
+
+        // TODO: how to calculate usdcDeposit
+        uint256 minUsdcDeposit = (_endTimestamp - block.timestamp) * _sizeLimit * SECRET_STORE_FEE_RATE * NO_OF_NODES_TO_SELECT;
+        _checkUsdcDeposit(_usdcDeposit, minUsdcDeposit);
+
+        USDC_TOKEN.safeTransferFrom(_owner, address(this), _usdcDeposit);
+
+        SelectedEnclave[] memory selectedEnclaves = SECRET_STORE.selectEnclaves(NO_OF_NODES_TO_SELECT, _sizeLimit);
+        if (selectedEnclaves.length < NO_OF_NODES_TO_SELECT)
+            revert SecretManagerUnavailableResources();
+
+        uint256 id = ++secretId;
+        userStorage[id].owner = _owner;
+        userStorage[id].sizeLimit = _sizeLimit;
+        userStorage[id].usdcDeposit = _usdcDeposit;
+        userStorage[id].startTimestamp = block.timestamp;
+        userStorage[id].endTimestamp = _endTimestamp;
+        userStorage[id].lastUsdcUpdateTimestamp = block.timestamp;
```

Comment:
we should start the billing for user secret after all acks are received from stores, not the creation time.